### PR TITLE
check in linter for required python run-time dependency

### PR DESCRIPTION
**Pull request by kain88-de**
This will ensure that conda is forced to add a python version dependency on the
final package.

I encountered this problem for the MDAnalysis feedstock. 
https://github.com/conda-forge/mdanalysis-feedstock/issues/4

I haven't tested this locally yet because I don't know if I can go back to my a stable build or even run the tests without installing the development version.

### CODE REVIEW

**Comment by kain88-de** [comment]
test pass now on travis

File: conda_smithy/lint_recipe.py [hunk 1]
```diff
@@ -149,6 +149,22 @@ def lintify(meta, recipe_dir=None):
...
+        lints.append("When using cython then python should be a run-time dependency")
+
+
+    # 14: If ``skip: True  # [py3k]`` python should be a run-time dependency
+    if recipe_dir is not None and os.path.exists(meta_fname):
+        with io.open(meta_fname, 'rt') as fh:
+            for selector_line in selector_lines(fh):
+                if 'py3k' in selector_line and 'python' not in requirements_section.get('run', ''):
+                    lints.append("When building python 2 only packages include python as a run-time "
+                                 "dependency to ensure it can't be installed on python 3.")
```
Comment (isuruf) [inline conda_smithy/lint_recipe.py]: I have some packages that needs python as a build dependency, but does not require python. I use one python because I don't need a matrix of builds. In that case, according to this, it's a failure

File: conda_smithy/lint_recipe.py [hunk 2]
```diff
@@ -149,6 +149,22 @@ def lintify(meta, recipe_dir=None):
     except RuntimeError as e:
         lints.append(str(e))
 
+    # 13: If cython is a build-dependency then python should be a run-time
+    # dependency. This ensures that packages will be marked correctly by conda
+    if 'cython' in requirements_section.get('build', '') and 'python' not in requirements_section.get('run', ''):
+        lints.append("When using cython then python should be a run-time dependency")
```
Comment (isuruf) [inline conda_smithy/lint_recipe.py]: This seems very specific. There are number of errors like this, but we can't add all of them. (swig, numpy as build dependency.)

How about something more generic like if python is an implicit runtime requirement, then it should be added explicitly in runtime requirements?

Comment (kain88-de) [inline conda_smithy/lint_recipe.py]: How would you detect an implicit requirement?

Comment (kain88-de) [inline conda_smithy/lint_recipe.py]: Extending this also to numpy/swig/etc isn't hard

File: conda_smithy/lint_recipe.py (hunk 1 above)
Comment (kain88-de) [inline conda_smithy/lint_recipe.py]: This only ensures a requirement  for Python if the package is Python 2 only. could you give an example problem where this would be a problem.

Comment (isuruf) [inline conda_smithy/lint_recipe.py]: For example, I have a package with a build script written in python 2. So, I skip py3k, but the package itself doesn't need python in run.

Comment (isuruf) [inline conda_smithy/lint_recipe.py]: eg: https://github.com/conda-forge/libcxx-feedstock/blob/master/recipe/meta.yaml

Comment (kain88-de) [inline conda_smithy/lint_recipe.py]: Maybe have this just for packages from pypi?

File: conda_smithy/lint_recipe.py [hunk 3]
```diff
@@ -149,6 +149,25 @@ def lintify(meta, recipe_dir=None):
     except RuntimeError as e:
         lints.append(str(e))
 
+    # 13: If cython is a build-dependency then python should be a run-time
+    # dependency. This ensures that packages will be marked correctly by conda
+    build_deps = ['cython', 'numpy x.x', 'swig']
```
Comment (kain88-de) [inline conda_smithy/lint_recipe.py]: Should there also be a lint that we check numpy is specified as `numpy x.x` when it appears in the build requirements?

File: conda_smithy/lint_recipe.py (hunk 1 above)
Comment (ocefpaf) [inline conda_smithy/lint_recipe.py]: This cannot be made into a lint error unless we can inform the linter that it is a python package.

File: conda_smithy/lint_recipe.py (hunk 3 above)
Comment (jakirkham) [inline conda_smithy/lint_recipe.py]: `swig` can be used for lots of things (not just `python`). Would rather not require people have `python` with `swig`.

File: conda_smithy/lint_recipe.py (hunk 1 above)
Comment (kain88-de) [inline conda_smithy/lint_recipe.py]: therefore my suggestion for packages from pypi. The assumption there is of course that pypi has only python packages.

Comment (ocefpaf) [inline conda_smithy/lint_recipe.py]: @kain88-de that assumption is blurred with wheels and we have tons of package that are python but not hosted at `PyPI` making this solution "flaky." So using the source URL to determined that is not reliable.

TL;DR unless we add an extra metadata to indicate that the recipe is pertinent to a python package we cannot do this.

Comment (kain88-de) [inline conda_smithy/lint_recipe.py]: OK then I'll drop this in this PR. But I have another question could we have 2 warning types in the linter. The first would be errors that we have now. The second type would be suggestions/hints to help packagers.

Comment (jakirkham) [inline conda_smithy/lint_recipe.py]: Maybe [namespaces]( https://gist.github.com/mcg1969/da5aec380d2ed083b79ddcf151ca16f1 ) could be leveraged depending on if/when/how they are constructed.

File: conda_smithy/lint_recipe.py (hunk 3 above)
Comment (kain88-de) [inline conda_smithy/lint_recipe.py]: I removed swig from that list. Should I add the `numpy x.x` check?

File: conda_smithy/lint_recipe.py (hunk 1 above)
Comment (jakirkham) [inline conda_smithy/lint_recipe.py]: > The second type would be suggestions/hints to help packagers.

Indeed. This sort of idea has come up before ( https://github.com/conda-forge/conda-smithy/issues/317 ). Some warnings about this might need `python` could be a reasonable compromise IMHO.

Comment (ocefpaf) [inline conda_smithy/lint_recipe.py]: > OK then I'll drop this in this PR. But I have another question could we have 2 warning types in the linter. The first would be errors that we have now. The second type would be suggestions/hints to help packagers.

Indeed warning type linting is something we should do!

File: conda_smithy/lint_recipe.py (hunk 3 above)
Comment (jakirkham) [inline conda_smithy/lint_recipe.py]: IMHO that seems ok.

File: conda_smithy/lint_recipe.py [hunk 4]
```diff
@@ -149,6 +149,22 @@ def lintify(meta, recipe_dir=None):
...
 
+    # 13: If cython/numpy is a build-dependency then python should be a run-time
+    # dependency. This ensures that packages will be marked correctly by conda
+    build_deps = ['cython', 'numpy x.x']
+    need_python_run_time = any([dep in requirements_section.get('build', '') for dep in build_deps])
+    if need_python_run_time and 'python' not in requirements_section.get('run', ''):
+        lints.append("Found python related compilation requirement "
+                     "please add python as a run-time dependency")
+
+    # 14: If numpy is in build requirements ensure it's stated as `numpy x.x`
```
Comment (kain88-de) [inline conda_smithy/lint_recipe.py]: here is the additional check for numpy x.x in the build requirements. The tests pass now again.

**Comment by kain88-de** [comment]
I removed the numpy check again. With https://github.com/conda-forge/conda-forge.github.io/pull/415 it changed anyway how to setup numpy

**Comment by kain88-de** [comment]
Is there still any interest in this?

**Comment by kain88-de** [comment]
@ocefpaf still interested in lint check?

**Comment by ocefpaf** [comment]
We are moving away from the `numpy x.x` syntax, so I guess that this PR may be DOA.

I'll leave the decision to you and the conda-smithy authors.

**Comment by kain88-de** [comment]
Is there still interest in this?  I think updating this to the new default of building against old numpy versions shouldn’t be to difficult.

**Comment by kain88-de** [comment]
Is this still interesting or can the PR be closed?
